### PR TITLE
[Rework] JAMES-2599 Metrics for BlobStores

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

File: server/blob/blob-memory/src/main/java/org/apache/james/blob/memory/MemoryBlobStore.java [hunk 1]
```diff
@@ -28,50 +28,79 @@
...
-        return new ByteArrayInputStream(retrieveStoredValue(blobId));
+    public MetricFactory metricFactory() {
+        return metricFactory;
     }
 
-    private byte[] retrieveStoredValue(BlobId blobId) {
-        return blobs.getOrDefault(blobId, new byte[]{});
+    @Override
+    public String metricNamePrefix() {
+        return "MemoryBlobStore:";
```
Comment (chibenwa) [inline server/blob/blob-memory/src/main/java/org/apache/james/blob/memory/MemoryBlobStore.java]: The 'metric prefix' should be the same for all blob stores IMO.

Comment (trantienduchn) [reply server/blob/blob-memory/src/main/java/org/apache/james/blob/memory/MemoryBlobStore.java]: we don't want to distinguish BlobStore impl in grafana board?

**Comment by mbaechler** [comment]
It's not what I meant. I described composition and you implemented inheritance.
The design I propose is : 
```java
class MetricBlobStore implements BlobStore {
  final BlobStore delegate;
  final MetricsFactory metrics;
  MetricBlobStore(MetricsFactory metrics, BlobStore delegate) {
    this.delegate = delegate;
    this.metrics = metrics
  }

  CompletableFuture<byte[]> readBytes(BlobId blobId) {
    return runWithMetrics(() -> delegate.readBytes(blobId));
  }

}
```
Then you always bind MetricBlobStore as BlobStore and inject the implementation in it.

**Comment by trantienduchn** [comment]
```
[ed49b2f1439925478bacbbb66bd6daaf45aba275] [ERROR] Errors: 
[ed49b2f1439925478bacbbb66bd6daaf45aba275] [ERROR]   ObjectStorageBlobStoreModuleTest.shouldSetupBlobStore:125 » Configuration Guic...
[ed49b2f1439925478bacbbb66bd6daaf45aba275] [ERROR]   ObjectStorageBlobStoreModuleTest.shouldSetupBlobStore:125 » Configuration Guic...
[ed49b2f1439925478bacbbb66bd6daaf45aba275] [ERROR]   ObjectStorageBlobStoreModuleTest.shouldSetupBlobStore:125 » Configuration Guic...
[ed49b2f1439925478bacbbb66bd6daaf45aba275] [INFO] 
[ed49b2f1439925478bacbbb66bd6daaf45aba275] [ERROR] Tests run: 43, Failures: 0, Errors: 3, Skipped: 0
```

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**